File: ts-core/src/models/query-order.ts
/**
 * Unified payment status values returned by the SDK.
 *
 * These normalise both query-order (`UPPER_CASE`) and webhook (`camelCase`)
 * trade statuses into a single set of values.
 */
export enum PaymentStatus {
  /** Payment completed successfully. */
  SUCCESS = 'SUCCESS',
  /** Payment failed. */
  FAIL = 'FAIL',
  /** Order expired or was closed without payment. */
  TIMEOUT = 'TIMEOUT',
  /** Awaiting payment or payment in progress. */
  PENDING = 'PENDING',
  /** Order accepted, awaiting processing. */
  ACCEPTED = 'ACCEPTED',
  /** Refund is in progress. */
  REFUNDING = 'REFUNDING',
  /** Refund completed successfully. */
  REFUND_SUCCESS = 'REFUND_SUCCESS',
  /** Refund failed. */
  REFUND_FAILED = 'REFUND_FAILED',
}

/** Maps raw Telebirr query-order `trade_status` values to the unified {@link PaymentStatus}. */
const TELEBIRR_STATUS_MAP: Record<string, PaymentStatus> = {
  PAY_SUCCESS: PaymentStatus.SUCCESS,
  PAY_FAILED: PaymentStatus.FAIL,
  ORDER_CLOSED: PaymentStatus.TIMEOUT,
  WAIT_PAY: PaymentStatus.PENDING,
  PAYING: PaymentStatus.PENDING,
  ACCEPTED: PaymentStatus.ACCEPTED,
  REFUNDING: PaymentStatus.REFUNDING,
  REFUND_SUCCESS: PaymentStatus.REFUND_SUCCESS,
  REFUND_FAILED: PaymentStatus.REFUND_FAILED,
};

/**
 * Maps a raw Telebirr `trade_status` string to the unified {@link PaymentStatus} enum.
 *
 * Unrecognised statuses default to `PENDING`.
 *
 * @param rawStatus - The raw `trade_status` value from the Telebirr API.
 * @returns The corresponding {@link PaymentStatus}.
 *
 * @example
 * ```ts
 * mapTelebirrStatus('PAY_SUCCESS'); // PaymentStatus.SUCCESS
 * mapTelebirrStatus('UNKNOWN');     // PaymentStatus.PENDING
 * ```
 */
export function mapTelebirrStatus(rawStatus: string): PaymentStatus {
  return TELEBIRR_STATUS_MAP[rawStatus] ?? PaymentStatus.PENDING;
}

/**
 * Parameters for querying an existing Telebirr order.
 */
export interface QueryOrderParams {
  /** The merchant order ID originally passed when creating the order. */
  merchOrderId: string;
}

/**
 * Response from a query-order API call.
 */
export interface QueryOrderResponse {
  /** Telebirr response code (`'0'` indicates success). */
  code: string;

  /** Normalised payment status. */
  status: PaymentStatus;

  /** Raw JSON response from the Telebirr API for debugging or advanced use. */
  rawResponse: Record<string, unknown>;
}

File: dart_core/lib/src/models/query_order.dart
/// The unified payment status returned by the SDK.
///
/// This enum normalizes the different status strings from Telebirr's
/// query-order and webhook APIs into a single set of values.
///
/// | Telebirr Query Status | Webhook Status | SDK Status       |
/// |-----------------------|----------------|------------------|
/// | `PAY_SUCCESS`         | `Completed`    | [success]        |
/// | `PAY_FAILED`          | `Failure`      | [fail]           |
/// | `ORDER_CLOSED`        | `Expired`      | [timeout]        |
/// | `WAIT_PAY`            | `Pending`      | [pending]        |
/// | `PAYING`              | `Paying`       | [pending]        |
/// | —                     | —              | [accepted]       |
/// | —                     | —              | [refunding]      |
/// | —                     | —              | [refundSuccess]  |
/// | —                     | —              | [refundFailed]   |
enum PaymentStatus {
  /// Payment completed successfully.
  success,

  /// Payment failed.
  fail,

  /// Order expired or was closed without payment.
  timeout,

  /// Payment is pending or in progress.
  pending,

  /// Order accepted by Telebirr, awaiting further processing.
  accepted,

  /// A refund is currently in progress.
  refunding,

  /// Refund completed successfully.
  refundSuccess,

  /// Refund failed.
  refundFailed,
}

/// Internal mapping from Telebirr's raw status strings to [PaymentStatus].
const _telebirrStatusMap = {
  'PAY_SUCCESS': PaymentStatus.success,
  'PAY_FAILED': PaymentStatus.fail,
  'ORDER_CLOSED': PaymentStatus.timeout,
  'WAIT_PAY': PaymentStatus.pending,
  'PAYING': PaymentStatus.pending,
  'ACCEPTED': PaymentStatus.accepted,
  'REFUNDING': PaymentStatus.refunding,
  'REFUND_SUCCESS': PaymentStatus.refundSuccess,
  'REFUND_FAILED': PaymentStatus.refundFailed,
};

/// Maps a raw Telebirr status string to a [PaymentStatus] enum value.
///
/// Unknown status strings default to [PaymentStatus.pending].
///
/// [rawStatus] is the `trade_status` value from the Telebirr API response
/// (typically in UPPER_CASE like `'PAY_SUCCESS'`).
///
/// Returns the corresponding [PaymentStatus] value.
///
/// Example:
/// ```dart
/// final status = mapTelebirrStatus('PAY_SUCCESS');
/// assert(status == PaymentStatus.success);
/// ```
PaymentStatus mapTelebirrStatus(String rawStatus) {
  return _telebirrStatusMap[rawStatus] ?? PaymentStatus.pending;
}

/// Parameters for querying an existing order's status.
///
/// Example:
/// ```dart
/// final params = QueryOrderParams(merchOrderId: 'ORDER_12345');
/// final response = await payments.query(params);
/// print('Status: ${response.status}');
/// ```
class QueryOrderParams {
  /// The merchant order ID to query.
  ///
  /// This is the same `merch_order_id` that was used when creating the order.
  final String merchOrderId;

  /// Creates a [QueryOrderParams] for the given [merchOrderId].
  const QueryOrderParams({required this.merchOrderId});
}

/// The response from a query-order API call.
///
/// Contains the mapped [status] and the full raw response for inspection.
///
/// Example:
/// ```dart
/// final response = await payments.query(QueryOrderParams(merchOrderId: '123'));
/// if (response.status == PaymentStatus.success) {
///   print('Payment completed!');
/// }
/// ```
class QueryOrderResponse {
  /// The Telebirr API response code.
  ///
  /// `'0'` indicates a successful query; other values may indicate errors.
  final String code;

  /// The mapped payment status.
  ///
  /// Transformed from the raw `trade_status` string using [mapTelebirrStatus].
  final PaymentStatus status;

  /// The full raw response map from the Telebirr API.
  ///
  /// Useful for accessing fields not explicitly modeled, such as `trade_status`
  /// in its original string form.
  final Map<String, dynamic> rawResponse;

  /// Creates a [QueryOrderResponse] with the query result.
  const QueryOrderResponse({
    required this.code,
    required this.status,
    required this.rawResponse,
  });
}
